csharp/span_based_serialization.md

```diff
@@ -0,0 +1,63 @@
+# More efficient Protobuf C# serialization/deserialization API
+
+## Background
+Currently serialization of protobuf C# messages is done (much like java) through instances of CodedInputStream and CodedOutputStream.
+These APIs work well, but there is some inherent overhead due to the fact that they are basically `byte[]` based.
+Recently a set of new types has been added to the .NET Framework (Span, ReadOnlySequence, IBufferWriter)
+and these new types allow defining new serialization/deserialization APIs, ones that would be much more efficient especially from
+the perspective of memory management.
+
+Both gRPC C# implementations (Grpc.Core and grpc-dotnet) already contain functionality that allows exposing request/response payloads
+in terms of the new .NET Span APIs in a way that requires no copying of binary payloads and basically allows deserialization directly
+from the buffer segments that were received from the network (and serializing directly into the buffer that is going to be sent out
+over the network).
+
+## Objective
+The goals for this effort:
+- keep the existing CodedInputStream and CodedOutputStream (no plans to remove them in the future), serialization/deserialization speed of these APIs should stay unaffected.
+- all the changes made need to be backward compatible
+- introduce a set of new public APIs for serialization/deserialization that utilize the Span type
+- allow future incremental performance optimizations and make them as easy as possible
+- ensure the are microbenchmark that allow comparing the two implementations
+- as support for the new serialization/deserialization APIs requires generating code that is not compatible with some older .NET frameworks, figure out a schema under which messages with "new serialization API support" and without it can coexist (without causing user friction).
+- the effort to maintain the Protobuf C# project should remain approximately the same (should not be negatively affected by introduction of new serialization/deserialization APIs).
+
+Non-goals
+- over-optimize the initial version of new serializer/deserializer (we should rather focus on enabling future improvements)
+- remove or deprecate the existing serialization/deserialization API
+
+## The Proposal
+
+Key elements:
+- Add CodedInputReader and CodedOutputWriter types (both are `ref structs` and thus can hold an instance of `Span`). These two types implement the low-level wire format semantics in terms of the `Span` type (as opposed to `byte[]` used by CodedInputStream and CodedOutputStream).
+- Introduce `IBufferMessage` interface (exact name TBD) which inherits from `IMessage` and marks protobuf messages that support the new serialization/deserialization API.
```

Comment:
I feel that the "Buffer"   in `IBufferMessage`, `PROTOBUF_DISABLE_BUFFER_SERIALIZATION`, `use_buffer_serialization` isn't that descriptive, so I'd still like to brainstorm other names for the feature.

Btw, what's the reasoning behind the name "Buffer"? Is the old parsing API considered "stream-based" which kind of makes the new API "buffer-based"? My concern is that this logic behind the naming isn't immediately obvious. 

Other ideas for names:
LOW_OVERHEAD_SERIALIZATION (I quite like that one, it makes a good self-explanatory option name, and an ok interface name `ILowOverheadSerializationMessage`)

PIPELINE_SERIALIZATION (don't like it that much but "Pipelines" is the name often used for the new IO style: 
https://devblogs.microsoft.com/dotnet/system-io-pipelines-high-performance-io-in-net/

Reply:
I chose Buffer because were writing to IBufferWriter and reading from ReadonlySequence, which is typically backed by a buffer memory pool. Also these types are found in the System.Buffers package.